akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/EffectfulActorContext.scala

```diff
@@ -83,4 +83,35 @@ import akka.annotation.InternalApi
     effectQueue.offer(Scheduled(delay, target, message))
     super.scheduleOnce(delay, target, message)
   }
+
+  override def mkTimer(): TimerScheduler[T] = new TimerScheduler[T] {
+    var activeTimers: Set[Any] = Set.empty
+    override def startTimerWithFixedDelay(key: Any, msg: T, delay: FiniteDuration): Unit =
+      startTimer(key, msg, delay, Effect.TimerScheduled.FixedDelayMode)
+
+    override def startTimerAtFixedRate(key: Any, msg: T, interval: FiniteDuration): Unit =
+      startTimer(key, msg, interval, Effect.TimerScheduled.FixedRateMode)
+
+    override def startPeriodicTimer(key: Any, msg: T, interval: FiniteDuration): Unit =
+      startTimer(key, msg, interval, Effect.TimerScheduled.FixedRateMode)
+
+    override def startSingleTimer(key: Any, msg: T, delay: FiniteDuration): Unit =
+      startTimer(key, msg, delay, Effect.TimerScheduled.SingleMode)
+
+    override def isTimerActive(key: Any): Boolean = ???
+
+    override def cancel(key: Any): Unit = if (activeTimers(key)) {
+      val effect = Effect.TimerCancelled(key)
+      effectQueue.offer(effect)
+      activeTimers -= key
+    }
+
+    override def cancelAll(): Unit = activeTimers.foreach(cancel)
+
+    def startTimer(key: Any, msg: T, delay: FiniteDuration, mode: Effect.TimerScheduled.TimerMode) = {
+      val effect = Effect.TimerScheduled(key, msg, delay, mode)
+      activeTimers += key
```

Comment:
In the actual impl starting a timer with an existing key replaces it, I wonder if we should perhaps capture that with a boolean in the effect as well given that we anyway know the key since before.